rfd/0078-role-variables.md

```diff
@@ -0,0 +1,176 @@
+        - match: '^alice@goteleport\.com$'
+          out: [root, admin]
+        # all teleport employees should be able to login as the local part of their email
+        - match: '@goteleport\.com$'
+          # "out" values can contain transforms and any traits
+          out: ["{{email.local(external.email)}}"]
+    - name: "allow_env"
+      # If the trait in a "select" expands to a list, all values will be checked
+      # for matches and the outputs will be combined.
+      select: "{{external.groups}}"
+      expansions:
+        # matchers can contain regexp capture groups, which can be expanded in
+        # "out" values
+        - match: '^env-(\w+)$'
+          out: ['$1']
+        - match: '^admins$'
+          out: ["prod"]
+  allow:
+    logins:
+      - "{{vars.logins}}"
+    node_labels:
+      env: ["{{vars.allow_env}}"]
+```
+
+These variables could be used in another role without redefinition:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: app-users
+spec:
+  allow:
+    app_labels:
+      env: ["{{vars.allow_env}}"]
+```
+
+### Should we select the first matching expansion, or all matching expansions?
+
+For the example
+
+```yaml
+vars:
+  - name: "logins"
+    select: "{{external.username}}"
+    expansions:
+      - match: "nic"
+        out: [root]
+      - match: "bob"
+        out: [readonly]
+      - match: "(.*)"
+        out: ["$1"]
+```
+
+we could either choose to expand the first match or all matches, the results for
+which are shown in this table:
+
+username | first match | all matches
+---------|-------------|------------
+nic      | [root]      | [root, nic]
+bob      | []          | [bob]
+alice    | [alice]     | [alice]
+
+In favor of expanding all matches, it is useful to be able to provide a default
+which applies to all possible values.
+
+In favor of selecting only the first match, it is useful to be able to exclude
+some value from matching any other expansions.
+This is especially true given Go's lack of support for negative-lookahead
+regular expressions.
+
+I'm not quite sure which is the best route to take here and am quite open to
+suggestions from reviewers.
+
+### When to compute variable values
+
+Variables will be computed during `FetchRoles`, before traits normally are
+expanded in the role definition.
+This first occurs during login, and again when building an `AccessChecker` on
+the proxy or on teleport Nodes/services before evaluating whether the user can
+access a resource.
+
+If it is desirable to avoid computing the variables more than once on login, the
+expanded values could be stored in the user's traits, which would be encoded in
+their certificates.
+However, this would make it difficult to detect variable redefinition and would
+make root cluster variables override alternate definitions in leaf clusters.
+
+### What happens when the same variable name is defined in two different roles?
+
+This is fine, until a user tries to login with both roles, in which case the
```

Comment:
That was going to be my suggestion as well.

Reply:
Yeah I like this idea.

Reply:
hmm, big downside here is that we use to Go parser which does not understand hyphens in the middle of identifiers, it thinks you are trying to do math with a minus sign. So roles like `node-users` wouldn't work, you'd have to rename to `node_users`...

Reply:
made this change

Reply:
Not sure if the parser would accept it, but maybe we can use string literals as a workaround? E.g. `vars."node-users".allow_env`, or something in that vein. Not great, but maybe it will allow to reference roles with dashes?

Reply:
I realized that the index syntax also works for this out of the box, i.e. you can do `{{var["role-name"]["variable-name"]}}` as necessary. Expanded on this in the doc.